File: app/src/main/java/com/witype/Dragger/mvp/activity/MainActivity.java
package com.witype.Dragger.mvp.activity;

import android.app.AlertDialog;
import android.content.DialogInterface;

import com.witype.Dragger.R;
import com.witype.Dragger.di.component.AppComponent;
import com.witype.Dragger.di.component.DaggerHomeComponent;
import com.witype.Dragger.mvp.BaseActivity;
import com.witype.Dragger.mvp.contract.HomeView;
import com.witype.Dragger.mvp.present.HomePresenter;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.BindView;

//import com.witype.Dragger.di.component.DaggerHomeComponent;

//import com.witype.Dragger.mvp.present.IHomePresenter;

public class MainActivity extends BaseActivity<HomePresenter> implements HomeView , SwipeRefreshLayout.OnRefreshListener {

    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;
    @BindView(R.id.refresh)
    SwipeRefreshLayout swipeRefreshLayout;

    @Inject
    QuarterAdapter quarterAdapter;

    AlertDialog alertDialog;

    @Override
    protected void setupComponent(@NonNull AppComponent appComponent) {
        DaggerHomeComponent.builder().view(this).appComponent(appComponent).build().inject(this);
    }

    @Override
    public int getResId() {
        return R.layout.activity_main;
    }

    @Override
    public void initView() {
        super.initView();
        setupRecycler();
        onRefresh();
    }

    @Override
    public void onGetMobileDataUsage() {

    }

    @Override
    public void onGetDataError(String message) {
        if (alertDialog != null) {
            alertDialog.dismiss();
        }
        String msg = getString(R.string.tip_occurred);
        msg = String.format("%s : %s",msg,message);
        alertDialog = new AlertDialog.Builder(this)
                .setMessage(msg)
                .setPositiveButton(R.string.dialog_case_retry, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        onRefresh();
                        dialogInterface.dismiss();
                    }
                })
                .create();
        alertDialog.show();
    }

    private void setupRecycler() {
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        swipeRefreshLayout.setOnRefreshListener(this);
        recyclerView.setAdapter(quarterAdapter);
    }

    @Override
    public void onRefresh() {
        getPresenter().getMobileDataUsage(100);
    }

    @Override
    public void showLoading() {
        //Activity 自己处理loading因此不需要调用父类的处理逻辑
//        super.showLoading();
        swipeRefreshLayout.setRefreshing(true);
    }

    @Override
    public void dismissLoading() {
//        super.dismissLoading();
        swipeRefreshLayout.setRefreshing(false);
    }

}

File: kotlindemo/src/main/java/com/witype/kotlindemo/mvp/ui/MainActivity.kt
package com.witype.kotlindemo.mvp.ui

import android.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import butterknife.BindView
import com.witype.kotlindemo.R
import com.witype.kotlindemo.di.component.DaggerHomeComponent
import com.witype.kotlindemo.mvp.contract.HomeView
import com.witype.kotlindemo.mvp.presenter.HomePresenter
import com.witype.mvp.di.component.AppComponent
import javax.inject.Inject

//import com.witype.Dragger.di.component.DaggerHomeComponent;
//import com.witype.Dragger.mvp.present.IHomePresenter;
class MainActivity : BaseActivity<HomePresenter>(), HomeView, OnRefreshListener {

    @BindView(R.id.recycler_view)
    lateinit var recyclerView: RecyclerView

    @BindView(R.id.refresh)
    lateinit var swipeRefreshLayout: SwipeRefreshLayout

    @set:Inject
    lateinit var quarterAdapter: QuarterAdapter

    var alertDialog: AlertDialog? = null

    override fun setupComponent(appComponent: AppComponent) {
        DaggerHomeComponent.builder().view(this).appComponent(appComponent).build().inject(this);
    }

    override val resId: Int get() = R.layout.activity_main

    override fun initView() {
        super.initView()
        setupRecycler()
        onRefresh()
    }

    override fun onGetMobileDataUsage() {}

    override fun onGetDataError(message: String?) {
        alertDialog?.dismiss()
        var msg = getString(R.string.tip_occurred)
        msg = String.format("%s : %s", msg, message)
        alertDialog = AlertDialog.Builder(this)
                .setMessage(msg)
                .setPositiveButton(R.string.dialog_case_retry) { dialogInterface, _ ->
                    onRefresh()
                    dialogInterface.dismiss()
                }
                .create()
        alertDialog?.show()
    }

    private fun setupRecycler() {
        val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)
        recyclerView.layoutManager = layoutManager
        swipeRefreshLayout.setOnRefreshListener(this)
        recyclerView.adapter = quarterAdapter
    }

    override fun onRefresh() {
        presenter!!.getMobileDataUsage(100)
    }

    override fun showLoading() {
        //Activity 自己处理loading因此不需要调用父类的处理逻辑
//        super.showLoading();
        swipeRefreshLayout.isRefreshing = true
    }

    override fun dismissLoading() {
//        super.dismissLoading();
        swipeRefreshLayout.isRefreshing = false
    }
}